modules/renter/downloadchunk.go

```diff
@@ -0,0 +1,149 @@
+	workersRemaining  int       // Number of workers still able to fetch the chunk.
+	workersStandby    []*worker // Set of workers that are able to work on this download, but are not needed unless other workers fail.
+
+	// Memory management variables.
+	memoryAllocated uint64
+
+	// The download object, mostly to update download progress.
+	download *download
+	mu       sync.Mutex
+}
+
+// fail will set the chunk status to failed. The physical chunk memory will be
+// wiped and any allocation will be returned to the renter. The download as a
+// whole will be failed as well.
+func (udc *unfinishedDownloadChunk) fail(err error) {
+	udc.failed = true
+	udc.recoveryComplete = true
+	for i := range udc.physicalChunkData {
+		udc.physicalChunkData[i] = nil
+	}
+	udc.download.managedFail(fmt.Errorf("chunk %v failed", udc.staticChunkIndex))
+}
+
+// threadedRecoverLogicalData will take all of the pieces that have been
+// downloaded and encode them into the logical data which is then written to the
+// underlying writer for the download.
+func (udc *unfinishedDownloadChunk) threadedRecoverLogicalData() error {
+	// Decrypt the chunk pieces.
+	udc.mu.Lock()
+	for i := range udc.physicalChunkData {
+		// Skip empty pieces.
+		if udc.physicalChunkData[i] == nil {
+			continue
+		}
+
+		key := deriveKey(udc.masterKey, udc.staticChunkIndex, uint64(i))
+		decryptedPiece, err := key.DecryptBytes(udc.physicalChunkData[i])
+		if err != nil {
+			udc.fail(err)
+			udc.mu.Unlock()
+			return err
+		}
+		udc.physicalChunkData[i] = decryptedPiece
+	}
+
+	// Recover the pieces into the logical chunk data.
+	recoverWriter := new(bytes.Buffer)
+	err := udc.erasureCode.Recover(udc.physicalChunkData, udc.staticChunkSize, recoverWriter)
+	if err != nil {
+		udc.fail(err)
+		udc.mu.Unlock()
+		return errors.AddContext(err, "unable to recover chunk")
+	}
+	// Clear out the physical chunk pieces, we do not need them anymore.
+	for i := range udc.physicalChunkData {
+		udc.physicalChunkData[i] = nil
+	}
+	udc.mu.Unlock()
+
+	// Write the bytes to the requested output.
+	start := udc.staticFetchOffset
+	end := udc.staticFetchOffset + udc.staticFetchLength
+	_, err = udc.destination.WriteAt(recoverWriter.Bytes()[start:end], udc.staticWriteOffset)
+	if err != nil {
+		udc.fail(err)
+		return errors.AddContext(err, "unable to write to download destination")
+	}
+	recoverWriter = nil
+
+	// Now that the download has completed and been flushed from memory, we can
+	// release the memory that was used to store the data. Call 'cleanUp' to
+	// trigger the memory cleanup along with some extra checks that everything
+	// is consistent.
+	udc.mu.Lock()
+	udc.recoveryComplete = true
+	udc.cleanUp()
+	udc.mu.Unlock()
+
+	// Update the download and signal completion of this chunk.
+	udc.download.mu.Lock()
+	defer udc.download.mu.Unlock()
+	udc.download.chunksRemaining--
+	atomic.AddUint64(&udc.download.atomicDataReceived, udc.staticFetchLength)
```

Comment:
`atomicDataReceived` probably doesn't need to be atomic if we always hold the `download.mu` when we access it.